add polyseed support

akildemir opened the pull request:
add support for polyseed to monero coin. Implementation taken from https://github.com/tevador/polyseed/.

Commit:
add polyseed support

kayabaNerve commented:
Clippy failure is unrelated.

kayabaNerve commented on coins/monero/src/tests/seed.rs:
```diff
@@ -6,27 +6,27 @@ use curve25519_dalek::scalar::Scalar;
   let vectors = [
     Vector {
-      language: Language::Chinese,
+      language: classic::Language::ChineseSimp,
```
Please spell out Simplified.

kayabaNerve commented on coins/monero/src/tests/seed.rs:
```diff
@@ -163,15 +163,113 @@ fn test_classic_seed() {
+    language: polyseed::Language,
+    seed: String,
+    key: String,
+    birtday: u64,
```
Mispelling of birthday.

kayabaNerve commented on coins/monero/src/tests/seed.rs:
```diff
@@ -163,15 +163,113 @@ fn test_classic_seed() {
+    Vector {
+      language: polyseed::Language::English,
+      seed: "raven tail swear infant grief assist regular lamp \
+      duck valid someone little harsh puppy airport language "
```
Can we automate testing with non-standard whitespace?

kayabaNerve commented on coins/monero/src/tests/seed.rs:
```diff
@@ -163,15 +163,113 @@ fn test_classic_seed() {
+    },
+    Vector {
+      language: polyseed::Language::Spanish,
+      seed: "eje fin part cele tabu pest lien puma \
```
Can we automate testing trimmed seeds?

kayabaNerve commented on coins/monero/src/tests/seed.rs:
```diff
@@ -163,15 +163,113 @@ fn test_classic_seed() {
+      birtday: 3118651200,
+      formal: false,
+    },
+  ];
```
We need vectors for all languages. Feather Wallet should be usable for them?

kayabaNerve commented on coins/monero/src/tests/seed.rs:
```diff
@@ -163,15 +163,113 @@ fn test_classic_seed() {
+    assert!(data.birthday <= v.birtday);
+    assert!(data.birthday + polyseed_time_step > v.birtday);
+
+    if v.formal {
```
I ran into this when testing trimmed classic seeds. The seed parts should always trim and then expand, where possible, in order to always generate the full string value.

kayabaNerve commented on coins/monero/src/wallet/seed/mod.rs:
```diff
@@ -21,44 +24,41 @@ pub enum SeedError {
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum SeedType {
+  Classic(classic::Language),
+  PolySeed(polyseed::Language),
```
Polyseed has a lower case "s".

kayabaNerve commented on coins/monero/src/wallet/seed/mod.rs:
```diff
@@ -67,26 +67,52 @@ impl Seed {
   /// Create a Seed from entropy.
-  pub fn from_entropy(lang: Language, entropy: Zeroizing<[u8; 32]>) -> Option<Seed> {
-    ClassicSeed::from_entropy(lang, entropy).map(Seed::Classic)
+  pub fn from_entropy(seed_type: SeedType, entropy: Zeroizing<[u8; 32]>) -> Option<Seed> {
```
My intent for the entropy function was for it to return the bytes usable as a key (it being a rename of spend_key which is somewhat inaccurate under Polyseed). I now realize I've messed up in that the entropy *is* defined as the original 19-bytes?

We should add a key function to return the generated *key*. This would be the entropy for classic seeds yet the PBKDF output for Polyseed.

kayabaNerve commented on coins/monero/src/wallet/seed/mod.rs:
```diff
@@ -67,26 +67,52 @@ impl Seed {
+  }
+
+  /// Return the polyseed data for this seed.
+  /// Return 0 values if it is a classic seed.
```
This should return an Option, not 0.

I'd also rather not have "polyseed_data", yet a dedicated birthday/features function. The birthday for a Classic seed should be 0 (not because it's unset but because no other value is safe). For features, it'd likely also be 0, signifying no features? Yet we may want to explicitly handle features in our API instead of relying on a uint expected to be interpreted per Polyseed. This may mean features returns Vec<Feature> or similar.

akildemir replied:
the problem with having separate birthday/features function is that it is inefficient. We will need to re-decode the seed per call. In one scenario, user needs all polyseed data and makes 1 call to `polyseed_data`, in another user makes 3 call to get the data they want and we will re-decode the same seed 3 times, although all the data was already available after the first call.

kayabaNerve commented on coins/monero/src/wallet/seed/mod.rs:
```diff
@@ -67,26 +67,52 @@ impl Seed {
+
+  /// Create a Seed from given data.
+  pub fn from_polyseed_data(data: PolySeedData, lang: polyseed::Language) -> Option<Seed> {
+    PolySeed::from_polyseed_data(data, lang).map(Seed::PolySeed)
```
Is it safe for the user to provide the checksum with this function?

akildemir replied:
I think so because checksum is ignored and always overwrited according to other field values.

akildemir replied:
function is converted into from(..args) now

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+  Czech,
+  Portuguese,
+  ChineseSimp,
+  ChineseTrad,
```
Please spell these out.

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+
+lazy_static! {
+  static ref LANGUAGES: HashMap<Language, WordList> = HashMap::from([
+    (Language::Czech, WordList::new(serde_json::from_str(CS_WORDS).unwrap(), true, false)),
```
You can have WordList::new perform this conversion to de-duplicate this.

akildemir replied:
you mean `serde_json::from_str(CS_WORDS).unwrap()` this conversion?

kayabaNerve replied:
Yep!

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+  result
+}
+
+fn poly_check(poly: &Poly) -> bool {
```
Does this really have to be a function when it only has one caller?

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+    let mut key = random_scalar(rng).to_bytes();
+    key[(SECRET_SIZE + 1) ..].fill(0);
+    key[SECRET_SIZE - 1] &= CLEAR_MASK;
+    let entropy = Zeroizing::new(key);
```
```rs
let entropy = Zeroizing::new([0; 32]);
rng.fill_bytes(entropy.as_mut());
```

(and then the fill/clear) :)

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+    let mut data = PolySeedData { birthday, features: 0, entropy, checksum: 0 };
+
+    // set the  checksum
+    data.checksum = create_checksum(&data);
```
Add a constructor for PolyseedData from these parts so the checksum can't be unset?

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+    let birthday = birthday_encode(data.birthday);
+
+    // make sure it is a valid scalar
+    Scalar::from_canonical_bytes(*data.entropy)?;
```
... why is this here? Anything which was created with

```rs
    let mut key = random_scalar(rng).to_bytes();
    key[(SECRET_SIZE + 1) ..].fill(0);
    key[SECRET_SIZE - 1] &= CLEAR_MASK;
    let entropy = Zeroizing::new(key);
```

will match this (as that's a 19-byte value needing to fit into a 31.5-byte space).

The concern I have is the non-19-byte values fitting into that 31.5 space.

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+    let data = poly_decode(&words)?;
+
+    // Make sure this is a valid scalar
+    let mut scalar = Scalar::from_canonical_bytes(*data.entropy);
```
See prior comment on Scalar validity/the entropy's expected form.

kayabaNerve commented on coins/monero/src/wallet/seed/polyseed.rs:
```diff
@@ -0,0 +1,419 @@
+    data
+  }
+
+  pub fn from(data: PolySeedData) -> Option<PolySeedData> {
```
This really should be new(features, birthday, entropy). PolyseedData -> Option<PolyseedData> suggests not all PolyseedDatas are valid. If they're invalid, they shouldn't be creatable.

Commit:
fix pr comments

Commit:
fix tests

kayabaNerve committed:
Embed the mempool into the Blockchain

kayabaNerve committed:
Plan scheduled payments whenever outputs are received

The scheduler prior waited for the next series of payments to be added.

kayabaNerve committed:
Replace Tendermint step with sync_block

Step moved a step forward after an externally synced/added block. This created
a race condition to add the block between the sync process and the Tendermint
machine. Now that the block routes through Tendermint, there is no such race
condition.

kayabaNerve committed:
Finish binding Tendermint into Tributary and define a Tributary master object

kayabaNerve committed:
Add correction the last commit missed

kayabaNerve committed:
Add DoS limits to tributary and require provided transactions be ordered

kayabaNerve committed:
Fix the scheduler from dropping UTXOs when there weren't any payments

kayabaNerve committed:
Documentation and cargo update

kayabaNerve committed:
Add a dedicated db crate with a basic DB trait

It's needed by the processor and tributary (coordinator).

kayabaNerve committed:
Add a DB to Tributary

Adds support for reloading most of the blockchain.

kayabaNerve committed:
Reloaded provided transactions from the disk

Also resolves a race condition by asserting provided transactions must be
unique, allowing them to be safely provided multiple times.

kayabaNerve committed:
must_use annotations on DbTxn

kayabaNerve committed:
Support reloading the mempool from disk

kayabaNerve committed:
Add a NewSet event to validator-sets

Updates to the latest serai-dex/substrate due to depending on
10ccaca0eb498a2316bbf627d419b29b1a75933a.

kayabaNerve committed:
Add basic getters to tributary

kayabaNerve committed:
cargo update

kayabaNerve committed:
Update to the latest subxt

Writes a custom unsigned extrinic creator due to subxt having an internal error
with the scale metadata. While the code in our scope increased, it's much more
ergonomic to our usage. We may end up rewriting most of subxt, eventually.

kayabaNerve committed:
Make unsigned private due to unsafe calling potential

kayabaNerve committed:
Start defining the coordinator

kayabaNerve committed:
Merge AckBlock with Burns

Offers greater efficiency while reducing concerns re: atomicity.

kayabaNerve committed:
Correct processor flow to have the coordinator decide signing set/re-attempts

The signing set should be the first group to submit preprocesses to Tributary.
Re-attempts shouldn't be once every 30s, yet n blocks since the last relevant
message.

Removes the use of an async task/channel in the signer (and Substrate signer).
Also removes the need to be able to get the time from a coin's block, which was
a fragile system marked with a TODO already.

kayabaNerve committed:
cargo +nightly fmt

kayabaNerve committed:
cargo update

Since p256 now pulls in an extra crate with this update, the {k,p}256 imports
disable default-features to prevent growing the tree.

kayabaNerve committed:
Support extracting timestamps from blocks

kayabaNerve committed:
Make progres on handling NewSet events

Further bones out the coordinator.

kayabaNerve committed:
Resolve #245

kayabaNerve committed:
Have InInstructions track the latest block for a network in storage

kayabaNerve committed:
Fill out code for the rest of the Substrate events

kayabaNerve committed:
Clean up the Substrate block processing code

kayabaNerve committed:
Rename transaction file to tributary, add function for genesis

kayabaNerve committed:
Add a processor API to the coordinator

kayabaNerve committed:
Add extensive commentary on mutable to the processor's main file

Clearly establishes why consistency is guaranteed from a Rust borrow-checker
mindset. While there are plenty of... 'violations', they're clearly explained.

Hopefully, this method of thinking helps promote/ensure consistency in the
future.

kayabaNerve committed:
Move ConfirmKeyPair from key_gen to substrate

Clarifies the emitter and accordingly why its mutations are justified.

kayabaNerve committed:
Remove BatchSigned

SubstrateBlock's provision of the most recently acknowledged block has
equivalent information with the same latency. Accordingly, there's no need for
it.

kayabaNerve committed:
Add note to processor_messages

kayabaNerve committed:
Use a single txn for an entire coordinator message

Removes direct DB accesses whre possible. Documents the safety of the rest.
Does uncover one case of unsafety not previously noted.

kayabaNerve committed:
cargo update to remove usage of yanked crate

kayabaNerve committed:
Clarify safety of Scanner::block_number and KeyGen::keys

kayabaNerve committed:
Tweak ConfirmKeyPair to alleviate database requirements of coordinator

kayabaNerve committed:
Use an enum for Coin/NetworkId

It originally wasn't an enum so software which had yet to update before an
integration wouldn't error (as now enums are strictly typed). The strict typing
is preferable though.

kayabaNerve committed:
Code a method to determine the activation block before any block has consensus

[0; 32] is a magic for no block has been set yet due to this being the first
key pair. If [0; 32] is the latest finalized block, the processor determines
an activation block based on timestamps.

This doesn't use an Option for ergonomic reasons.

Commit:
automate whitespace & trimming test cases

kayabaNerve committed:
Save keys by their tweaked group_key

Keys are referred to by their tweaked versions. If a tweak was needed, keys
would fail to confirm.

kayabaNerve committed:
Use crypto-bigint's reduction in ed448

Achieves feasible performance in the ed448 which makes it potentially viable
for real world usage.

Accordingly prepares a new release, updating the README.

kayabaNerve committed:
Move the entirety of ed448 to Residue, offering a further 2-4x speedup

kayabaNerve committed:
Resolve #68

Notably speeds up monero-serai's build and CLSAG performance.

kayabaNerve committed:
Make MainDB into SubstrateDB

kayabaNerve committed:
Initial Tributary handling

kayabaNerve committed:
Add additional checks to key_gen/sign

There is the ability to cause state bloat by flooding Tributary.
KeyGen/Sign specifically shouldn't allow bloat since we check the
commitments/preprocesses/shares for validity. Accordingly, any invalid data
(such as bloat) should be detected.

It was posssible to place bloat after the valid data. Doing so would be
considered a valid KeyGen/Sign message, yet could add up to 50k kB per sign.

kayabaNerve committed:
Apply DKG TX handling code to all sign TXs

The existing code was almost entirely applicable. It just needed to be scoped
with an ID. While the handle function is now a bit convoluted, I don't see a
better option.

kayabaNerve committed:
Split FinalizedBlock into ExternalBlock and SeraiBlock

Also re-arranges their orders.

kayabaNerve committed:
Add support for multiple orderings in Provided

Necessary as our Tributary chains needed to agree when a Serai block has
occurred, and when a Monero block has occurred. Since those could happen at the
same time, some validators may put SeraiBlock before ExternalBlock and vice
versa, causing a chain halt. Now they can have distinct ordering queues.

kayabaNerve committed:
Slash on unrecognized ID

kayabaNerve committed:
ExternalBlock handler

kayabaNerve committed:
Add a SubstrateBlockAck message to the processor

When a Substrate block occurs, the coordinator is expected to emit
SubstrateBlock. This causes the processor to begin a variety of plans. The
processor now emits SubstrateBlockAck, explicitly listing all plan IDs, before
starting signing.

This lets the coordinator provide a SubstrateBlock transaction, and with it,
recognize all plan IDs as valid.

Prior, we would've had to have a spotty algorithm based upon the upcoming
Preprocess messages, or if we immediately provided the SubstrateBlock
transaction, then wait for the processor to inform us of the contained plans.

This creates an explicitly proper async flow not reliant on waiting for data
availability.

Alternatively, we could've replaced Preprocess with (Block, Vec<Preprocess>).
This would've been more efficient, yet also clunky due to the multiple usages
of the Preprocess message.

kayabaNerve committed:
Route the SubstrateBlock message, which is the last Tributary transaction type

kayabaNerve committed:
Add recent bloat checks added to signer to substrate_signer as well

kayabaNerve committed:
Add no_std support to transcript, dalek-ff-group, ed448, ciphersuite, multiexp, schnorr, and monero-generators

transcript, dalek-ff-group, ed449, and ciphersuite are all usable with no_std
alone. The rest additionally require alloc.

Part of #279.

kayabaNerve committed:
Add a test to the coordinator for running a Tributary

Impls a LocalP2p for testing.

Moves rebroadcasting into Tendermint, since it's what knows if a message is
fully valid + original.

Removes TributarySpec::validators() HashMap, as its non-determinism caused
different instances to have different round robin schedules. It was already
prior moved to a Vec for this issue, so I'm unsure why this remnant existed.

Also renames the GH no-std workflow from the prior commit.

kayabaNerve committed:
Add a test for Tributary

Further fleshes out the Tributary testing code.

kayabaNerve committed:
Test handling of DKG commitments transactions

kayabaNerve committed:
Add Transaction::sign.

While I don't love the introduction of empty_signed, it's practically fine.

kayabaNerve committed:
Tributary test wait_for_tx_inclusion function

kayabaNerve committed:
Additionally test DKGShares

kayabaNerve committed:
Handle adding new Tributaries

Removes last_block as an argument from Tendermint. It now loads from the DB as
needed. While slightly less performant, it's easiest and should be fine.

kayabaNerve committed:
Reload Tributaries

add_active_tributary writes the spec to disk before it returns, so even if the
VecDeque it pushes to isn't popped, the tributary will still be loaded on boot.

kayabaNerve committed:
Start handling P2P messages

This defines the tart of a very complex series of locks I'm really unhappy
with. At the same time, there's not immediately a better solution. This also
should work without issue.

kayabaNerve committed:
Clarify Arc RwLocks and sleeps in coordinator

kayabaNerve committed:
Send a heartbeat message when a Tributary falls behind

kayabaNerve committed:
cargo fmt

kayabaNerve committed:
cargo update

kayabaNerve committed:
Move json word lists to rs

Allows building the seed code without serde_json.

kayabaNerve committed:
Break coordinator main into multiple functions

Also moves from std::sync::RwLock to tokio::sync::RwLock to prevent wasting
cycles on spinning.

kayabaNerve committed:
Remove reliance on a blockchain read lock from block/commit

kayabaNerve committed:
Implement Tributary syncing

Also adds a forwards-lookup to the Tributary blockchain.

kayabaNerve committed:
Don't return from sync_block until the Tendermint machine returns if it's valid or not

We had a race condition where'd we be informed of blocks 1 .. 3, and
immediately add 1 .. 3. Because we immediately tried to add 2 after 1, it'd
fail since the tip was still the genesis, yet 2 needs the tip to be 1.

Adding a channel, while ugly, was the simplest way to accomplish this.

Also has any added block be broadcasted. Else there's a race condition where a
node which syncs up to the most recent block does so, yet fails to add the next
block when it's committed to.

kayabaNerve committed:
Test handle_p2p and Tributary syncing

Includes bug fixes.

kayabaNerve committed:
Tweak tests workflow

kayabaNerve committed:
Add a TributaryReader which doesn't require a borrow to operate

Reduces lock contention.

Additionally changes block_key to include the genesis. While not technically
needed, the lack of genesis introduced a side effect where any Tributary on the
the database could return the block of any other Tributary. While that wasn't a
security issue, returning it suggested it was on-chain when it wasn't. This may
have been usable to create issues.

kayabaNerve committed:
Document panic in FROST

kayabaNerve committed:
Document a pair of panics requiring 256 GB of RAM/4 GB of a context

kayabaNerve committed:
Add a UID function to messages

When we receive messages, we're provided with a message ID we can use to
prevent handling an item multiple times. That doesn't prevent us from *sending*
an item multiple times though. Thanks to the UID system, we can now not send if
already present.

Alternatively, we can remove the ordered message ID for just the UID, allowing
duplicates to be sent without issue, and handled on the receiving end.

kayabaNerve committed:
Initial code to handle messages from processors

kayabaNerve committed:
Document the processor/tributary/coordinator/serai flow

kayabaNerve committed:
Have Coordinator MainDb take a mutable borrow

kayabaNerve committed:
Update to substrate polkadot-v0.9.42

kayabaNerve committed:
Correct error message in ff-group-tests

kayabaNerve committed:
Update to May's nightly

Doesn't use the PR due to the needed changes.

kayabaNerve committed:
Support arbitrary RPC providers in monero-serai

Sets a clean path for no-std premised RPCs (buffers to an external RPC impl)/
Tor-based RPCs/client-side load balancing/...

kayabaNerve committed:
Correct processor's handling of the new Monero RPC code

kayabaNerve committed:
Correct Serai Dockerfile

kayabaNerve committed:
Publish ExternablBlock/SubstrateBlock, delay *Preprocess until ID acknowledged

Adds a channel for the Tributary scanner to communicate when an ID has been
acknowledged.

kayabaNerve committed:
Rename uid to intent

kayabaNerve committed:
Use U448 for Ed448 instead of U512

kayabaNerve committed:
Spawn a new async task for each block message

This probably should be done with n-long lived tasks, one per Tributary. While
this may not be suitably performant long-term (potential DoS vector), this at
least resolves the halting concerns.

kayabaNerve committed:
Move the coordinator to a n-processor design

kayabaNerve committed:
Ensure Tributary commits are minimal

kayabaNerve committed:
Properly get genesis for a Processor message

kayabaNerve committed:
Create a vote transaction upon GeneratedKeyPair

kayabaNerve committed:
Remove TODO about code de-duplication

It's infeasible to write a macro/function there. Does add a type alias which
makes things cleaner.

kayabaNerve committed:
Have coordinator publish batches to Substrate

kayabaNerve committed:
Implement MuSig key aggregation into DKG

Isn't spec compliant due to the lack of a spec to be compliant too.

Slight deviation from the paper by using a unique list instead of a multiset.

Closes #186, progresses #277.

kayabaNerve committed:
Correct 2/3rds definitions throughout the codebase

The prior formula failed for some values, such as 20.
20 / 3 = 6, * 2 = 12, + 1 = 13. 13 is 65%, not >= 67.

kayabaNerve committed:
cargo update

Resolves a yanked crate and removes some duplicated dependencies.

kayabaNerve committed:
Add a dedicated function to get a MuSig key

kayabaNerve committed:
Do the minimal amount of work for dkg to compile under no-std

The Substrate runtime requires access to the MuSig key aggregation function.

\#279 related.

kayabaNerve committed:
Use a MuSig signature to publish validator set key pairs to Serai

The processor/coordinator flow still has to be rewritten.

kayabaNerve committed:
Correct various no_std definitions

kayabaNerve committed:
Add a context to MuSig key aggregation

kayabaNerve committed:
Use proper messages for ValidatorSets/InInstructions pallet

Provides a DST, and associated metadata as beneficial.

Also utilizes MuSig's context to session-bind. Since set_keys_messages also
binds to set, this is semi-redundant, yet that's appreciated.

kayabaNerve committed:
Remove signed Substrate TXs from Coordinator

kayabaNerve committed:
Only scan v2 Monero TXs

kayabaNerve committed:
Fix for prior commit

kayabaNerve committed:
Ensure canonical points in the cross-group DLEq proof

Commit:
Update nightly

kayabaNerve committed:
Fix incorrect sig_hash generation

sig_hash was used as a challenge. challenges should be of the form H(R, A, m).
These sig hashes were solely H(A, m), allowing trivial forgeries.

kayabaNerve committed:
cargo update

Resolves an openssl advisory and nets ~-8 crates.

kayabaNerve committed:
Build no-std tests with RISC-V 32 IMAC

Turns out wasm still has std, making it suboptimal to use here.

kayabaNerve committed:
Pin setup-protoc to v2.0.0

Commit:
fix tributary sync test

kayabaNerve committed:
Update to substrate polkadot-v0.9.43

kayabaNerve committed:
Slight terminology correction in sync test

Also correct a mistake from merging the most recent polkadot version.

kayabaNerve committed:
Replace lazy_static with OnceLock inside monero-serai

lazy_static, if no_std environments were used, effectively required always
using spin locks. This resolves the ergonomics of that while adopting Rust std
code.

no_std does still use a spin based solution. Theoretically, we could use
atomics, yet writing our own Mutex wasn't a priority.

kayabaNerve committed:
no-std support for monero-serai (#311)

* Move monero-serai from std to std-shims, where possible

* no-std fixes

* Make the HttpRpc its own feature, thiserror only on std

* Drop monero-rs's epee for a homegrown one

We only need it for a single function. While I tried jeffro's, it didn't work
out of the box, had three unimplemented!s, and is no where near viable for
no_std.

Fixes #182, though should be further tested.

* no-std monero-serai

* Allow base58-monero via git

* cargo fmt

kayabaNerve committed:
Represent RCT amounts with None, not 0.

Fixes #282.

Does allow any v1 TXs which exist, and v2 miner-TXs, to specify Some(0). As far
as I can tell, both were/are theoreitcally possible.

Commit:
Update nightly

kayabaNerve committed:
Add a message queue

This is intended to be a reliable transport between the processors and
coordinator. Since it'll be intranet only, it's written as never fail.

Primarily needs testing and a proper ID.

kayabaNerve committed:
cargo update

Resolves https://github.com/serai-dex/serai/security/dependabot/29

Boog900 committed:
std-shims: six `Read` for &[u8]

kayabaNerve committed:
Correct deny.toml with inclusion of message-queue

kayabaNerve committed:
Use serai- prefixes on Serai-specific packages

Fixes deny.toml, also runs a minor cargo update shrinking the tree.

kayabaNerve committed:
Update monero-tests workflow to new name for the processor

kayabaNerve committed:
Correct depends for processor-messages

kayabaNerve committed:
Disable Rust caching

We hit the cache limit after just one or two builds, making it infeasible.

kayabaNerve committed:
cargo update

Resolves a yanked crate

kayabaNerve committed:
Move location of serai-client in Cargo.toml

Boog900 committed:
Monero: support for legacy transactions (#308)

* add mlsag

* fix last commit

* fix miner v1 txs

* fix non-miner v1 txs

* add borromean + fix mlsag

* add block hash calculations

* fix for the jokester that added unreduced scalars

to the borromean signature of
2368d846e671bf79a1f84c6d3af9f0bfe296f043f50cf17ae5e485384a53707b

* Add Borromean range proof verifying functionality

* Add MLSAG verifying functionality

* fmt & clippy :)

* update MLSAG, ss2_elements will always be 2

* Add MgSig proving

* Tidy block.rs

* Tidy Borromean, fix bugs in last commit, replace todo! with unreachable!

* Mark legacy EcdhInfo amount decryption as experimental

* Correct comments

* Write a new impl of the merkle algorithm

This one tries to be understandable.

* Only pull in things only needed for experimental when experimental

* Stop caching the Monero block hash now in processor that we have Block::hash

* Corrections for recent processor commit

* Use a clearer algorithm for the merkle

Should also be more efficient due to not shifting as often.

* Tidy Mlsag

* Remove verify_rct_* from Mlsag

Both methods were ports from Monero, overtly specific without clear
documentation. They need to be added back in, with documentation, or included
in a node which provides the necessary further context for them to be naturally
understandable.

* Move mlsag/mod.rs to mlsag.rs

This should only be a folder if it has multiple files.

* Replace EcdhInfo terminology

The ECDH encrypted the amount, yet this struct contained the encrypted amount,
not some ECDH.

Also corrects the types on the original EcdhInfo struct.

* Correct handling of commitment masks when scanning

* Route read_array through read_raw_vec

* Misc lint

* Make a proper RctType enum

No longer caches RctType in the RctSignatures as well.

* Replace Vec<Bulletproofs> with Bulletproofs

Monero uses aggregated range proofs, so there's only ever one Bulletproof. This
is enforced with a consensus rule as well, making this safe.

As for why Monero uses a vec, it's probably due to the lack of variadic typing
used. Its effectively an Option for them, yet we don't need an Option since we
do have variadic typing (enums).

* Add necessary checks to Eventuality re: supported protocols

* Fix for block 202612 and fix merkel root calculations

* MLSAG (de)serialisation fix

ss_2_elements will not always be 2 as rct type 1 transactions are not enforced to have one input

* Revert "MLSAG (de)serialisation fix"

This reverts commit 5e710e0c96658092c6ecfe5e4ea5a9c3dbee3ab3.

here it checks number of MGs == number of inputs:
https://github.com/monero-project/monero/blob/0a1eaf26f9dd6b762c2582ee12603b2a4671c735/src/cryptonote_core/tx_verification_utils.cpp#L60-59

and here it checks for RctTypeFull number of MGs == 1:
https://github.com/monero-project/monero/blob/0a1eaf26f9dd6b762c2582ee12603b2a4671c735/src/ringct/rctSigs.cpp#L1325

so number of inputs == 1
so ss_2_elements == 2

* update `MlsagAggregate` comment

* cargo update

Resolves a yanked crate

* Move location of serai-client in Cargo.toml

---------

Co-authored-by: Luke Parker <lukeparker5132@gmail.com>

kayabaNerve committed:
Fix the known issue with the DSA

I wrote it to only select TXs with a timelock, not only TXs which are unlocked.
This most likely explains why it so heavily selected coinbases.

Also moves an InternalError which would've never been hit on mainnet, yet
technically isn't an invariant, to only exist when cfg(test).

kayabaNerve committed:
Add a bin to download a chain, over RPC, reserializing and hashing every item

Parallelized. Doesn't check the deserialization is correct. Does use distinct,
persistent HTTP clients.

kayabaNerve committed:
Correct how Monero integration tests are run

kayabaNerve committed:
Support multiple RPCs in the reserialize_chain bin

kayabaNerve committed:
Don't call get_height every block

kayabaNerve committed:
Modify get_transactions to split requests as to not hit the restricted RPC limits

kayabaNerve committed:
Meaningful changes from aggressive-clippy

I do want to enable a few specific lints, yet aggressive-clippy as a whole
isn't worthwhile.

kayabaNerve committed:
Extend reserialize_chain with CLSAG/BP(+) verification

kayabaNerve committed:
Remove spammy println from reserialize_chain

kayabaNerve committed:
Update reserialize_chain for v1 and migration TXs

Also always marks 0-amount inputs as RCT due to impossibility of non-RCT
0-amount outputs.

kayabaNerve committed:
Only deserialize RctSignatures where's there at least one input

This is only enforced by the Monero protocol due to a single check the mixRing
isn't empty in get_pre_mlsag_hash. The value in ensuring there's a least one
input is to ensure the safety of our rct_type functions, which determines the
RctType based off structural analysis (specifically, input data if
MlsagBorromean).

rct_type was technically safe without this. A 0-input transaction would be
mis-classified as RctFull/MlsagAggregate, which would then make the
RctSignatures invalid for being RctFull (requiring exactly one input) yet not
having inputs, meaning an invalid RctSignatures would be mis-classified yet
still invalid.

This just removes the risk of mis-classification in the first place, tightening
the library's safety.

Steven-Chang committed:
docs/Getting Started.md: cargo build --release --all-features

kayabaNerve committed:
Fix the known instance of #295

kayabaNerve committed:
Bind RocksDB into serai-db

kayabaNerve committed:
Split up tests in CI to avoid node storage limits

kayabaNerve committed:
Corrections to prior commit

kayabaNerve committed:
Again

I called git commit --amend without calling git add . again :(

kayabaNerve committed:
Update the flow for completed signing processes

Now, an on-chain transaction exists. This resolves some ambiguities and
provides greater coordination.

kayabaNerve committed:
Merge branch 'develop' of https://github.com/akildemir/serai into polyseed

kayabaNerve committed:
Clean Polyseed code

kayabaNerve committed:
Final tweaks

kayabaNerve committed:
Correct no-std builds for Polyseed

kayabaNerve committed:
Again correct no-std
